compiler/AST/symbol.cpp

```diff
@@ -1438,11 +1439,34 @@ void createInitStringLiterals() {
   stringLiteralModule->block->insertAtTail(new DefExpr(initStringLiterals));
 }
 
+bool isValidString(std::string str) {
+  return chpl_enc_validate_buf(str.c_str(), str.length()) == 0;
+}
+
 // Note that string immediate values are stored
 // with C escapes - that is newline is 2 chars \ n
 // so this function expects a string that could be in "" in C
 VarSymbol *new_StringSymbol(const char *str) {
 
+  // String (as record) literals are inserted from the very beginning on the
+  // parser all the way through resolution (postFold). Since resolution happens
+  // after normalization we need to insert everything in normalized form. We
+  // also need to disable parts of normalize from running on literals inserted
+  // at parse time.
+
+  // these are created first so that we can unescape string early for validation
+  VarSymbol* cstrTemp = newTemp("call_tmp");
+  CallExpr *cstrMove = new CallExpr(PRIM_MOVE, cstrTemp, new_CStringSymbol(str));
+
+  std::string unescapedString = unescapeString(str, cstrMove);
+
+  if (!isValidString(unescapedString)) {
+    USR_FATAL_CONT("Invalid string literal");
+  }
+  else {
+    // TODO: somehow flag that the buffer is validated to avoid revalidation
+  }
+
```

Comment:
I don't think you should move this VarSymbol creating code. You should just add it after the conditional `if (s)` below. The idea would be that the `stringLiteralsHash` would only contain validated UTF-8. But, for that to make sense, you'd need to create a sibling to `stringLiteralsHash`, e.g. `bytesLiteralsHash`, and use that in `new_BytesSymbol`.

But the bottom line is that I don't think it's a good idea to make this code do anything it doesn't absolutely have to before checking the hash. Otherwise we create problems for ourselves.